api/api-public-tenant-permissions.go

```diff
@@ -130,37 +103,164 @@ func (s *server) AddPermission(ctx context.Context, in *pb.AddPermissionRequest)
+	}
+	// TODO: parallelize
+	resourcesToCreate := differenceArrays(resourcesBucketNames, currentResourceBucketNames)
+	resourcesToDelete := differenceArrays(currentResourceBucketNames, resourcesBucketNames)
+	// CREATE New Resources
+	// create a Temporal permission to Create the new Permission resources
+	tempPermission := &cluster.Permission{ID: permission.ID}
+	cluster.AppendPermissionResourcesObj(tempPermission, resourcesToCreate)
+	// for each resource, save to DB
+	for _, resc := range tempPermission.Resources {
+		err = cluster.InsertResource(appCtx, tempPermission, &resc)
+		if err != nil {
+			return nil, err
+		}
+	}
+	// DELETE unwanted Resources
+	for _, bucketName := range resourcesToDelete {
+		resourceID, ok := mapResourceName[bucketName]
+		if !ok {
+			return nil, status.New(codes.Internal, "error retrieving resource to delete").Err()
+		}
+		err = cluster.DeletePermissionResourceDB(appCtx, resourceID)
+		if err != nil {
+			return nil, status.New(codes.Internal, "error deleting existing resource").Err()
 		}
-		res.Actions = append(res.Actions, &pbAction)
 	}
 
-	if permissionObj.Description != nil {
-		res.Description = *permissionObj.Description
+	// -- Update Permission ACTIONS
+	var currentActionTypes []string
+	mapActionTypes := make(map[string]string)
+	for _, perm := range permission.Actions {
+		currentActionTypes = append(currentActionTypes, string(perm.ActionType))
+		mapActionTypes[string(perm.ActionType)] = perm.ID.String()
 	}
-	return res
+	// TODO: parallelize
+	actionsToCreate := differenceArrays(actionTypes, currentActionTypes)
+	actionsToDelete := differenceArrays(currentActionTypes, actionTypes)
+	// CREATE New Actions
+	cluster.AppendPermissionActionObj(tempPermission, actionsToCreate)
+	// for each resource, save to DB
+	for _, action := range tempPermission.Actions {
+		err = cluster.InsertAction(appCtx, tempPermission, &action)
+		if err != nil {
+			return nil, err
+		}
+	}
+	// DELETE unwanted Actions
+	for _, types := range actionsToDelete {
+		actionID, ok := mapActionTypes[types]
+		if !ok {
+			return nil, status.New(codes.Internal, "error retrieving action to delete").Err()
+		}
+		err = cluster.DeletePermissionActionDB(appCtx, actionID)
+		if err != nil {
+			return nil, status.New(codes.Internal, "error deleting existing action").Err()
+		}
+	}
+
+	// Update single parameters
+	err = cluster.UpdatePermissionDB(appCtx, permission)
+	if err != nil {
+		return nil, status.New(codes.Internal, "error updating permission").Err()
+	}
+	// if we reach here, all is good, commit
+	if err := appCtx.Commit(); err != nil {
+		return nil, err
+	}
+
+	// get updated permission
+	updatedPermission, err := cluster.GetPermissionByID(appCtx, permission.ID.String())
+	if err != nil {
+		return nil, status.New(codes.InvalidArgument, "permission not found").Err()
+	}
+	// Create response object
+	permissionResponse := buildPermissionResponseFromPermissionObj(updatedPermission)
+
```

Comment:
lastly, you need to see which service accounts use this permission and update their policies:
```
// update the policy for each SA
	var saChs []chan error
	for _, sa := range serviceAccountIDs {
		ch := UpdatePolicyForServiceAccount(ctx, sgt.StorageGroupTenant, tenantConf, sa)
		saChs = append(saChs, ch)
	}
	// wait for all to finish
	for _, ch := range saChs {
		err := <-ch
		if err != nil {
			return err
		}
	}
```

Reply:
done please review